Snapshot

tan-yue opened the pull request:
1. add optional --load_from and --dump_to both of which take snapshot directory as argument
2. add optional --mem_size to set guest memory size
3. make rootfs and appfs writable otherwise applications cannot run.
4. add customized `kvm-bindings` which uses `serde`.

tan-yue committed:
add --from_file cmdline arg

tan-yue committed:
add customized kvm-bindings

tan-yue committed:
add optional --mem_size to set guest memory size

tan-yue committed:
make --load_from and --dump_to take directory that holds snapshot as
argument

tan-yue committed:
make rootfs and appfs writable

alevy commented:
For `kvm-bindings`, instead of putting the source code in _this_ repository, you should fork it to SNS, make appropriate changes in the forked repo and point our fork of firecracker to that repo rather.

This addresses two issues with the currently proposed setup:

1. Currently it's hard to see which changes you actually made to kvm-bindings, making it harder to upstream them if necessary (and also harder to review)

2. The current setup has an odd cross-dependency between this repo and the firecracker sub-repository, where the firecracker sub-repository can't compile on it's own, since the kvm-bindings dependency relies on firecracker living in a subdirectory of the firecracker-tools repo

tan-yue committed:
add optional --vcpu_count to set the number of vcpus

tan-yue committed:
remove kvm-bindings
point firecracker to the forked repo instead

tan-yue commented:
1. removed local `kvm-bindings` as suggested. [changes](https://github.com/princeton-sns/kvm-bindings/compare/master...princeton-sns:serde) made to kvm-bindings
2. also added optional --vcpu_count as I tested multi-vcpu snapshot.

alevy commented on firerunner/src/main.rs:
```diff
@@ -135,7 +187,7 @@ fn main() {
         drive_id: String::from("rootfs"),
         path_on_host: PathBuf::from(rootfs),
         is_root_device: true,
-        is_read_only: true,
```
Using a read only file system for both the rootfs and appfs works for me, and is preferable from a design perspective (we definitely don't want applications changing the shared root file system, and they also shouldn't change their own file system since each boot should start fresh).

Maybe this is a snapshot specific thing?

tan-yue replied:
I do not think it is snapshot specific. The error I came across was that kernel paniced when trying to write to /tmp/workload.pid but finding out that the file system is read only.

alevy replied:
Hrm... Well I don't see why the application needs to do that. My versions of the images don't do that. Besides of we want to write to tmp we should just mount it as a tmpfs (in memory) instead of part of the disk drive.

alevy commented on firerunner/src/main.rs:
```diff
@@ -39,12 +56,12 @@ fn main() {
                 .takes_value(true)
                 .required(false)
-                .default_value("console=ttyS0 reboot=k panic=1 pci=off")
+                .default_value("console=ttyS0 reboot=k panic=1 pci=off nokaslr")
```
Is `nokaslr` necessary? Is it relevant to creating snapshots?

tan-yue replied:
oops, no. I forgot to remove it.

alevy commented on firerunner/src/main.rs:
```diff
@@ -88,6 +121,8 @@ fn main() {
     let appfs = cmd_arguments.value_of("appfs");
     let cmd_line = cmd_arguments.value_of("command line").unwrap().to_string();
+    let mem_size = cmd_arguments.value_of("mem_size");
+    let vcpu_cnt = cmd_arguments.value_of("vcpu_cnt");
```
`let vcpu_count`

we're not limited to 60 character line lengths and variable names are compiled away, so there no performance hit to longer names. The same could be said for `cmd_line`, `rootfs`, `mem_size`, etc, but `cnt` instead of `count` is maybe a bridge too far :)

alevy commented on firerunner/src/main.rs:
```diff
@@ -109,8 +144,16 @@ fn main() {
     let shared_info = Arc::new(RwLock::new(InstanceInfo {
         state: InstanceState::Uninitialized,
         id: instance_id,
+        load_dir: None,
```
Instead of starting with `None` and writing later based on the if-branch, just:

```rust
    load_dir: cmd_arguments.value_of("load_dir").map(PathBuf::from),
    dump_dir: cmd_arguments.value_of("dump_dir").map(PathBuf::from),
```

alevy commented on firerunner/src/main.rs:
```diff
@@ -123,6 +166,15 @@ fn main() {
     };
 
+    let mut machine_config = VmConfig::default();
+    if let Some(mem_size) = mem_size {
```
Ditto here, you can just use `.map()` to avoid the if-branch:

```rust
machine_config.mem_size_mib = mem_size.map(|ms| ms.parse().unwrap())
```

alevy reviewed (changes requested):
In general, do all the parsing and validation of command line arguments first (e.g. converting strings into integers), then use them more simply in the application logic (e.g. when setting fields in shared_info etc)

tan-yue committed:
minor style changes; mark rootfs and appfs as read-only

LedgeDash commented:
@tan-yue 
This actually broke build on my end.
The error is at `firecracker-tools/firecracker/vmm/src/lib.rs:1328`. The commit is `1faee03a6b6e31691a1a401895cc1cd4554e1b39` to the firecracker repo. I also added a comment at the offending line in that commit.

It's due to some "borrowing `self `as mutable more than once at a time". I briefly read the rust reference, couldn't quite understand what's going on. So figured I'd report here and since one of you should be able to fix it much faster.

tan-yue commented:
@LedgeDash your firecracker repo should point to commit `b53fe76a210526d5d1d55a533857086306e5edb1` for firerunner to build.
The `master` branch of firecracker-tools now points at commit `b53fe76a210526d5d1d55a533857086306e5edb1`.

LedgeDash commented:
```
[~/Dev/firecracker-tools/firecracker|Sat Aug 10 17:16:21]
[davidliu@sns59] git show
commit b53fe76a210526d5d1d55a533857086306e5edb1 (HEAD, origin/snapshot, origin/HEAD, snapshot)
Author: Yue Tan <yuetan@cs.princeton.edu>
Date:   Fri Aug 9 21:34:27 2019 -0500

    add set_queues/get_queues for VsockEpollHandler
    so that firecracker can compile when `vsock` feature is on
```
Yeah, that's the one I'm at.

LedgeDash commented:
Problem resolved. toolchain issue. updating from 1.33 to 1.36 solved the problem. Will see if I can specify required rustc version in the `toml `file.

alevy commented:
Not in the toml file but in a special `rust_toolchain` file. See the Tock repo for an example
